src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/CSharpVirtualDocumentPublisher.cs

```diff
@@ -48,6 +48,19 @@ public override void Initialize(LSPDocumentManager documentManager)
         // Internal for testing
         internal void DocumentManager_Changed(object sender, LSPDocumentChangeEventArgs args)
         {
+            // We need the below check to address a race condition between when a request is sent to the C# server
+            // and when the C# server receives a workspace/didOpen notification for the document. This race condition
+            // may occur when the Razor server finishes initializing before C# receives and processes the document open
+            // request.
+            // This workaround adds the Razor client name to the document so the C# server will recognize it, despite
+            // the document not being formally opened. Note this is meant to only be a temporary workaround until a
+            // longer-term solution is implemented in the future.
+            if (args.Kind == LSPDocumentChangeKind.Added)
+            {
+                var csharpContainer = new CSharpVirtualDocumentContainer(_lspDocumentMappingProvider, args.New, args.New.Snapshot);
```

Comment:
Ah so my one concern with this approach is that when you're shifting from "closed" -> "opened" this will very briefly empty the C# document. I worry that may have unintended impacts cross-document where we may lose Blazor component information or get unexpected errors etc. For instance I'd imagine that if you had Index.razor opened and then opened SurveyPrompt.razor you may get the "Title" attribute in in the default Index.razor document as not being bound.

To address this we effectively need to mark the already-existing "background" document in the dynamic file info as "RazorLSP" so that the content that's already there exists but can be treated correctly. Let me know if I can add clarity

Reply:
Digging deeper, I still feel a little unclear about the correct approach here. Per our conversation earlier, it seemed like part of the ideal solution involves creating a `CSharpVirtualDocumentContainer` for the dynamic document. However, generating a `CSharpVirtualDocumentContainer` requires getting the dynamic document's LSP snapshot and ITextSnapshot, which I'm not sure how to retrieve from the present infrastructure. Any ideas on how to get the snapshots, or is there some other piece of the puzzle I'm missing? Trying to see if there's a document manager somewhere but no luck so far. :(

Reply:
Ah ya I'm sorry, I was definitely floundering a little bit in that conversation and I can totally see where the confusion set in. Sooo you're in the right direction. What you need is a hybrid of `CSharpVirtualDocumentContainer` and `DefaultDynamicDocumentContainer`. The `DefaultDynamicDocumentContainer` is what's [used today to update background document content](https://github.com/dotnet/aspnetcore-tooling/blob/c2a8e1d67e122aa8c3b936ccc2bb1ece6579831a/src/Razor/src/Microsoft.VisualStudio.Editor.Razor/BackgroundDocumentGenerator.cs#L298). Now that requires a `DocumentSnapshot` which is the inprocess representation of a Razor file; pretty much equivalent to Roslyn's `TextDocument`.

Now in your case you want evvvverything from the default dynamic continer EXCEPT the properties. Reason being, when those properties change you're basically handing off from the "background" Razor world to the "foreground" Razor world. IIRC the default dynamic document container is sealed so you may need to unseal it to accomplish this but that should at least give you a good starting point.

Now the last piece of the puzzle is how do we get a `DocumentSnapshot`. In Razor scenarios our "Workspace" is representative of a `ProjectSnapshotManager`. There are 2 of them in LSP scenarios. Both basically have the same content; however, the one that runs in VS' whole purpose is for the updating/tracking of background files and the one that runs in the LSP server manages basically allll documents but because it's separate from VS it can't update Roslyn's representation of our Razor files so it's primarily used for open documents. Sooo given the `CSharpVirtualDocumentPublisher` runs in-proc in VS (denoted by it being in the `Microsoft.VisualStudio..LanguageServerClient.Razor assembly) you can import the VS `ProjectSnapshotManager` in the ctor of the publisher and should be able to look at every project for the corresponding document, something like:
```C#
foreach (var project in _projectSnapshotManager.Projects)
{
    var documentUri = args.New.Uri;
    var filePath = args.New.Uri.GetAbsoluteOrUNCPath().Replace('/', '\\');
    var document = project.GetDocument(filePath);
    if (document != null)
    {
        .. You have a document! ..
```

Hopefully this adds a bit more clarity and direction!

Reply:
I added logic that hopefully promotes the background document correctly. Let me know if it needs tweaking!